### PR TITLE
API  from `ParquetExec` to `ParquetExecBuilder`

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion/issues/12737
Closes https://github.com/apache/datafusion/pull/12726

## Rationale for this change

See https://github.com/apache/datafusion/issues/12737 for rationale (basically many times we manipulate a `ParquetExec` downstream we end up needing to create a builder from an existing `ParquetExec`

## What changes are included in this PR?

Changes:
1. Add `ParquetExec::into_builder()` method
2. Add docs with examples of usage
3. Update some existing code to use it

## Are these changes tested?
Yes

## Are there any user-facing changes?
New API and Docs

### CODE REVIEW

**Comment by alamb** [comment]
fyi @NGA-TRAN

File: datafusion/core/src/datasource/physical_plan/parquet/mod.rs [hunk 1]
```diff
@@ -479,9 +544,15 @@ impl ParquetExec {
         self.pruning_predicate.as_ref()
     }
 
+    /// return the optional file reader factory
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/parquet/mod.rs:547]: I added this API for completeness

File: datafusion/core/src/datasource/physical_plan/parquet/mod.rs [hunk 2]
```diff
@@ -586,12 +662,8 @@ impl ParquetExec {
...
     }
 
-    fn with_file_groups(mut self, file_groups: Vec<Vec<PartitionedFile>>) -> Self {
-        self.base_config.file_groups = file_groups;
-        // Changing file groups may invalidate output partitioning. Update it also
-        let output_partitioning = Self::output_partitioning_helper(&self.base_config);
-        self.cache = self.cache.with_partitioning(output_partitioning);
-        self
+    fn with_file_groups(self, file_groups: Vec<Vec<PartitionedFile>>) -> Self {
+        self.into_builder().with_file_groups(file_groups).build()
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/parquet/mod.rs:666]: the builder forces the recalculation of the various properties -- I think the existing code actually has a bug that the statistics are not recomputed when new file groups are added

Comment (alamb) [reply datafusion/core/src/datasource/physical_plan/parquet/mod.rs]: update-- I was wrong. The existing code only uses the existing file groups so statistics don't need to be updated. I renamed this function to make this more clear

File: datafusion/core/src/datasource/physical_plan/parquet/mod.rs [hunk 3]
```diff
@@ -166,6 +166,33 @@ pub use writer::plan_to_parquet;
...
+///
+/// ```no_run
+/// # use std::sync::Arc;
+/// # use arrow::datatypes::Schema;
+/// # use datafusion::datasource::physical_plan::{FileScanConfig, ParquetExec};
+/// # use datafusion::datasource::listing::PartitionedFile;
+/// # fn parquet_exec() -> ParquetExec { unimplemented!() }
+/// // Split a single ParquetExec into multiple ParquetExecs, one for each file
+/// let exec = parquet_exec();
+/// let exiting_file_groups = &exec.base_config().file_groups;
```
Comment (NGA-TRAN) [inline datafusion/core/src/datasource/physical_plan/parquet/mod.rs:182]:
```suggestion
/// let existing_file_groups = &exec.base_config().file_groups;
```

File: datafusion/core/src/datasource/physical_plan/parquet/mod.rs [hunk 4]
```diff
@@ -166,6 +166,33 @@ pub use writer::plan_to_parquet;
...
+/// let new_execs = exiting_file_groups
```
Comment (NGA-TRAN) [inline datafusion/core/src/datasource/physical_plan/parquet/mod.rs:183]:
```suggestion
/// let new_execs = existing_file_groups
```

File: datafusion/core/src/datasource/physical_plan/parquet/mod.rs [hunk 5]
```diff
@@ -459,6 +498,32 @@ impl ParquetExec {
         ParquetExecBuilder::new(file_scan_config)
     }
 
+    /// Convert this `ParquetExec` into a builder for modification
+    pub fn into_builder(self) -> ParquetExecBuilder {
+        // list out fileds so it is clear what is being dropped
```
Comment (NGA-TRAN) [inline datafusion/core/src/datasource/physical_plan/parquet/mod.rs:503]:
```suggestion
        // list out fields so it is clear what is being dropped
```

Comment (NGA-TRAN) [reply datafusion/core/src/datasource/physical_plan/parquet/mod.rs:503]: I wonder if the dropped fields will be rebuilt during the `built()` function?

**Review by NGA-TRAN — APPROVED** [review]
Nice. Thanks @alamb

**Review by jayzhan211 — APPROVED** [review]
👍

File: datafusion/core/src/datasource/physical_plan/parquet/mod.rs (hunk 5 above)
Comment (alamb) [reply datafusion/core/src/datasource/physical_plan/parquet/mod.rs:503]: Yes they are -- I will make that clear in comments

**Comment by alamb** [comment]
Thanks again @jayzhan211  and @NGA-TRAN

**Pull request merged**